cibuildwheel/macos.py

```diff
@@ -286,6 +281,11 @@ def setup_python(
                 # we can do universal2 builds on macos 10.15, but we need to
                 # set ARCHFLAGS otherwise CPython sets it to `-arch x86_64`
                 env.setdefault("ARCHFLAGS", "-arch arm64 -arch x86_64")
```

Comment:
You need to set `_PYTHON_HOST_PLATFORM=macosx-11.0-universal2` and `ARCHFLAGS` here (regardless of the os version).

Reply:
Awesome, thanks! Though I think we could set `_PYTHON_HOST_PLATFORM=macosx-10.9-universal2`, since the Intel portion supports 10.9+ (it does require a slightly newer packaging/pip to be supported). Actually, should we get this number from value of `MACOSX_DEPLOYMENT_TARGET`?

Reply:
>  Actually, should we get this number from value of MACOSX_DEPLOYMENT_TARGET? 

Sure. (That might break a few workflows because the wheel tag will have say 10.11 instead of the expected 10.9 which users might have renamed to correctly reflect MACOS_DEPLOYMENT_TARGET. Though this should be minor and using MACOSX_DEPLOYMENT_TARGET is the correct strategy.)

Reply:
Setting this as the minimum (10.9) seems to do the right thing:

```
3 wheels produced in 13 minutes:
  boost_histogram-1.0.3.dev15+g7314060-cp38-cp38-macosx_10_12_universal2.whl
  boost_histogram-1.0.3.dev15+g7314060-cp38-cp38-macosx_10_12_x86_64.whl
  boost_histogram-1.0.3.dev15+g7314060-cp38-cp38-macosx_11_0_arm64.whl
```

Reply:
I went ahead & use MACOS_DEPLOYMENT_TARGET in the strategy. Feel free to drop the commit if not appropriate.

Reply:
Is there any benefit? The wheels were already correctly named, and there might be some protection if you set the deployment target too low, and this is a bit more, messier code.

Reply:
I find it messy to set `macosx_11_0_arm64` & `MACOSX_DEPLOYMENT_TARGET=10.9`. This raises questions because they're not synced. This way, they're synced and it makes sense when reading the code.
Maybe the code could be a bit less messy but at least it makes sense.

> there might be some protection if you set the deployment target too low

Can you expend a bit more on this ?

Reply:
> and this is a bit more, messier code.

Do you have a specific thing in mind or more of a general feeling ?

The former I can try to address. I'm not sure what to do about the latter.

Reply:
If you set `MACOSX_DEPLOYMENT_TARGET` (MDT) less than the `_PYTHON_HOST_PLATFORM`, you still get the correct wheel tag (the higher of the two). If you set it above, you also get the correct wheel tag (again, the higher of the two). Honestly, maybe it should even be an error to set MDT less than 10.9? Since it internally seems to be doing the right thing, I'm not sure if we need to "help" it. The more we take on, the more we are likely to break. We know the true value of `_PYTHON_HOST_PLATFORM` - I think it makes more sense to use that if Python is doing the right thing for us. If not, then yes, we could lie about it. (Now if there's a benefit, like if it was not producing the right wheel tags, it clearly would be better to fix it).

Does MDT need to be two digit anymore? The host will, for historical reasons, but not sure about MDT? On macOS 11, the patch version is now the minor version, so `MACOSX_DEPLOYMENT_TARGET=11` seems acceptable (though haven't tried it). The next version will be macOS 12, not 11.1 (we are already up to 11.4). This code seems to force it.

Reply:
> Does MDT need to be two digit anymore? The host will, for historical reasons, but not sure about MDT? On macOS 11, the patch version is now the minor version, so MACOSX_DEPLOYMENT_TARGET=11 seems acceptable (though haven't tried it). The next version will be macOS 12, not 11.1 (we are already up to 11.4). This code seems to force it.

Seems it does not need to be two digit anymore. I will update this.

```
pybase64 Matt$ MACOSX_DEPLOYMENT_TARGET=11 python setup.py bdist_wheel
.....
creating 'dist/pybase64-1.1.4-cp39-cp39-macosx_11_0_universal2.whl' and adding 'build/bdist.macosx-10.9-universal2/wheel' to it
....

pybase64 Matt$ ls dist
pybase64-1.1.4-cp39-cp39-macosx_11_0_universal2.whl
```

As a side note, scikit-build chokes on this:
```
cmake-python-distributions Matt$ MACOSX_DEPLOYMENT_TARGET=11 python setup.py bdist_wheel
Traceback (most recent call last):
  File "/Users/Matt/Dev/cmake-python-distributions/setup.py", line 7, in <module>
    from skbuild import setup
  File "/Users/Matt/Dev/py39/lib/python3.9/site-packages/skbuild/__init__.py", line 9, in <module>
    from .setuptools_wrap import setup  # noqa: F401
  File "/Users/Matt/Dev/py39/lib/python3.9/site-packages/skbuild/setuptools_wrap.py", line 39, in <module>
    from . import cmaker
  File "/Users/Matt/Dev/py39/lib/python3.9/site-packages/skbuild/cmaker.py", line 20, in <module>
    from .constants import (CMAKE_BUILD_DIR,
  File "/Users/Matt/Dev/py39/lib/python3.9/site-packages/skbuild/constants.py", line 39, in <module>
    _SKBUILD_PLAT_NAME = _default_skbuild_plat_name()
  File "/Users/Matt/Dev/py39/lib/python3.9/site-packages/skbuild/constants.py", line 34, in _default_skbuild_plat_name
    return 'macosx-{}.{}-{}'.format(split_ver[0], split_ver[1], machine)
IndexError: list index out of range
```

Reply:
> If you set MACOSX_DEPLOYMENT_TARGET (MDT) less than the _PYTHON_HOST_PLATFORM, you still get the correct wheel tag (the higher of the two). If you set it above, you also get the correct wheel tag (again, the higher of the two). Honestly, maybe it should even be an error to set MDT less than 10.9?

If you set it less than 10.9, at least on my system, the extension fails to build using regular setuptools.

> Since it internally seems to be doing the right thing, I'm not sure if we need to "help" it.

`wheel` seems to do the right thing in those cases so maybe it's not needed indeed. I tried to test various combination and all worked (or failed) as expected.

> The more we take on, the more we are likely to break. We know the true value of _PYTHON_HOST_PLATFORM - I think it makes more sense to use that if Python is doing the right thing for us. If not, then yes, we could lie about it. (Now if there's a benefit, like if it was not producing the right wheel tags, it clearly would be better to fix it).

You're most likely right here. There are too much risks compared to a tiny consistency issue I find when reading the code. feel free to drop the commits.

Reply:
Could you set CMAKE_OSX_ARCHITECTURES as well? https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html

Reply:
`scikit-build` (the official CMake build system for for Python) now supports `ARCHFLAGS`, and https://github.com/pybind/cmake_example does too. If you are wrapping CMake yourself, I think you should respect `ARCHFLAGS` as well. I don't think we can or should set every build tool's special flags for this, but the tooling should support the standard Python settings.

Reply:
See https://github.com/pybind/cmake_example/blob/f6539807e764707f2ee2b9d7ba0c8d017d685853/setup.py#L93-L97 , specifically, if you are not using Scikit-build for some reason.

Reply:
Good to know this. I will try Scikit-build. Would mind telling me who defined the "ARCHFLAGS" standard?

Reply:
It's from setuptools, and therefore probably from distutils in the Python standard library.